File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/telemetry/TelemetryConsent.kt
package ai.jolli.jollimemory.core.telemetry

/**
 * TelemetryConsent — opt-out gate for IntelliJ telemetry (JOLLI-1785 Phase 3).
 * Independent Kotlin implementation; matches the CLI/VS Code consent model so a
 * machine reads one consistent state. Telemetry is ON by default, silenced when
 * the user said no through any channel, in order of authority:
 *
 *   1. `DO_NOT_TRACK` env var (https://consoledonottrack.com): set & not "0".
 *   2. A host opt-out — the JetBrains data-sharing decline, when the startup
 *      wiring resolves it (passed in as `platformDisabled`; the consent core
 *      stays platform-agnostic and testable).
 *   3. The shared `telemetry: "off"` flag in ~/.jolli/jollimemory/config.json.
 */
object TelemetryConsent {
    enum class Reason { ON, DO_NOT_TRACK, PLATFORM_OFF, CONFIG_OFF }

    data class Result(val enabled: Boolean, val reason: Reason)

    /** Resolve the effective consent state and the reason behind it. */
    fun resolve(
        telemetryFlag: String?,
        env: Map<String, String> = System.getenv(),
        platformDisabled: Boolean = false,
    ): Result {
        if (doNotTrackSet(env)) return Result(false, Reason.DO_NOT_TRACK)
        if (platformDisabled) return Result(false, Reason.PLATFORM_OFF)
        if (telemetryFlag == "off") return Result(false, Reason.CONFIG_OFF)
        return Result(true, Reason.ON)
    }

    fun isEnabled(
        telemetryFlag: String?,
        env: Map<String, String> = System.getenv(),
        platformDisabled: Boolean = false,
    ): Boolean = resolve(telemetryFlag, env, platformDisabled).enabled

    /** Show the loud first-run notice once — only when enabled and not yet shown. */
    fun shouldShowNotice(
        noticeShown: Boolean,
        telemetryFlag: String?,
        env: Map<String, String> = System.getenv(),
        platformDisabled: Boolean = false,
    ): Boolean = !noticeShown && isEnabled(telemetryFlag, env, platformDisabled)

    private fun doNotTrackSet(env: Map<String, String>): Boolean {
        val v = env["DO_NOT_TRACK"]?.trim() ?: return false
        return v.isNotEmpty() && v != "0"
    }
}

File: cli/src/core/TelemetryConsent.ts
/**
 * TelemetryConsent — the single gate that decides whether telemetry may be
 * collected (JOLLI-1785 Phase 2). Opt-out model: telemetry is ON by default,
 * but is silenced when the user has said no through any of three channels,
 * checked in order of authority:
 *
 *   1. `DO_NOT_TRACK` — the cross-tool platform signal
 *      (https://consoledonottrack.com): set and not "0" ⇒ opt out.
 *   2. A host platform opt-out — VS Code passes
 *      `!vscode.env.isTelemetryEnabled` (which already folds in
 *      `telemetry.telemetryLevel === "off"`). This module deliberately takes
 *      it as a parameter rather than importing `vscode`, because CLI core is
 *      bundled verbatim into the extension and must not depend on the
 *      extension host.
 *   3. Our own config `telemetry: "off"` — the one-line in-app off switch.
 *
 * A loud first-run notice is shown once per machine (tracked by
 * `telemetryNoticeShown`); `shouldShowTelemetryNotice` decides when.
 */
import type { JolliMemoryConfig } from "../Types.js";

/** Why telemetry is on or off — surfaced by `jolli telemetry status|inspect`. */
export type ConsentReason = "on" | "do-not-track" | "platform-off" | "config-off";

export interface ConsentInput {
	/** The loaded global config (only the telemetry-related fields are read). */
	readonly config: Pick<JolliMemoryConfig, "telemetry" | "telemetryNoticeShown">;
	/** Environment to read `DO_NOT_TRACK` from. Defaults to `process.env`. */
	readonly env?: NodeJS.ProcessEnv;
	/**
	 * Host-platform opt-out. VS Code passes `!vscode.env.isTelemetryEnabled`.
	 * The CLI leaves this undefined (no host platform to honor).
	 */
	readonly platformDisabled?: boolean;
}

export interface ConsentResult {
	readonly enabled: boolean;
	readonly reason: ConsentReason;
}

/**
 * `DO_NOT_TRACK` is "set" for our purposes when present and not `"0"` and not
 * empty — matching the de-facto console DNT convention (any truthy value means
 * opt out; `"0"` explicitly means "tracking allowed").
 */
function doNotTrackSet(env: NodeJS.ProcessEnv): boolean {
	const raw = env.DO_NOT_TRACK;
	if (raw === undefined) return false;
	const v = raw.trim();
	return v !== "" && v !== "0";
}

/** Resolve the effective consent state and the reason behind it. */
export function resolveTelemetryConsent(input: ConsentInput): ConsentResult {
	const env = input.env ?? process.env;
	if (doNotTrackSet(env)) {
		return { enabled: false, reason: "do-not-track" };
	}
	if (input.platformDisabled === true) {
		return { enabled: false, reason: "platform-off" };
	}
	if (input.config.telemetry === "off") {
		return { enabled: false, reason: "config-off" };
	}
	return { enabled: true, reason: "on" };
}

/** Convenience boolean for the hot `track()` path. */
export function isTelemetryEnabled(input: ConsentInput): boolean {
	return resolveTelemetryConsent(input).enabled;
}

/**
 * The loud first-run notice is shown once — only when telemetry is actually
 * enabled (no point announcing collection we then won't do) and the notice has
 * not been recorded as shown yet. The caller persists `telemetryNoticeShown`
 * after displaying it.
 */
export function shouldShowTelemetryNotice(input: ConsentInput): boolean {
	if (input.config.telemetryNoticeShown === true) return false;
	return resolveTelemetryConsent(input).enabled;
}
